src/mon/MonClient.cc

```diff
@@ -111,14 +107,17 @@ int MonClient::get_monmap_privately()
 
   ldout(cct, 10) << "have " << monmap.epoch << " fsid " << monmap.fsid << dendl;
 
+  std::random_device rd;
```

Comment:
@jdurgin
```c++
#include <random>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

void print_rng(int n)
{
  vector<unsigned> ranks(n);
  for (unsigned i = 0; i < n; i++) {
    ranks[i] = i;
  }
  std::random_device rd;
  std::mt19937 rng(rd());
  std::shuffle(ranks.begin(), ranks.end(), rng);
  for (unsigned i = 0; i < n; i++) {
    cout << ranks[i] << ", ";
  }
  cout << endl;
}

int main()
{
  for (int i = 0; i < 10; i++) {
    print_rng(10);
  }
}
```
```
[kchai@rex001 tmp]$ ./a.out
2, 9, 7, 0, 6, 5, 8, 3, 1, 4,
6, 0, 9, 3, 2, 5, 1, 4, 7, 8,
6, 2, 7, 5, 4, 0, 3, 9, 8, 1,
2, 9, 0, 5, 8, 1, 3, 7, 6, 4,
5, 1, 9, 6, 2, 8, 3, 7, 4, 0,
7, 3, 8, 5, 1, 4, 0, 2, 6, 9,
2, 4, 9, 6, 8, 1, 0, 5, 7, 3,
4, 6, 3, 2, 5, 9, 0, 1, 8, 7,
9, 4, 1, 8, 0, 2, 3, 6, 7, 5,
4, 7, 2, 1, 0, 8, 5, 6, 9, 3,
[kchai@rex001 tmp]$ ./a.out
8, 7, 2, 5, 4, 3, 0, 1, 9, 6,
4, 1, 8, 6, 5, 9, 2, 3, 0, 7,
6, 7, 2, 9, 0, 8, 1, 5, 3, 4,
7, 1, 0, 8, 5, 4, 2, 9, 6, 3,
0, 9, 1, 6, 5, 2, 4, 7, 8, 3,
5, 2, 8, 4, 6, 9, 0, 7, 1, 3,
9, 3, 6, 0, 1, 7, 5, 4, 2, 8,
4, 9, 6, 5, 8, 0, 3, 1, 2, 7,
0, 2, 7, 3, 5, 1, 9, 4, 6, 8,
6, 3, 9, 0, 4, 1, 5, 2, 7, 8,
[kchai@rex001 tmp]$ gcc --version
gcc (GCC) 4.8.3 20140911 (Red Hat 4.8.3-7)
Copyright (C) 2013 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

also, i tested this with gcc 5.4.1, gcc 6.3, clang 3.9 and clang 4.0,  all of them are generating different sequences in each iteration in the `for` loop, and in each run. so `std::random_device` is able to generate **non-deterministic** random number sequence.

the article you are referencing states that

> C++11 random number generators can be initialized either with a single integer or with an object that matches the C++11 Seed Sequence concept.

right. i am using a single integer returned by `std::random_device::operator()` for creating this "single integer". and this "single integer" is used to seed the `mt19937` random number engine. let's look at how "random" the `std::random_device` is. `std::random_device` uses the `rdrand` instruction if available. see https://en.wikipedia.org/wiki/RdRand#Compilers. this matches with my finding testing with different versions of compilers.

> It Might Actually Be Deterministic

yes, it's scaring. but the point is "Might". right, libstdc++ always returns `0.0`. maybe it just cannot tell the entropy of `rdrand` instruction. the stackoverflow post referenced by that article applies to the gcc on MingW, which is not supported by us.